src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc

```diff
@@ -102,41 +193,283 @@ const std::array<VkLayerProperties, 1> kLayerProperties{{{
+  // generated from it.
+
+  GlobalData* global_data = GetGlobalData();
+  DeviceData* device_data =
+      global_data->device_map.get(device_key(queue))->get();
+
+  // Go through each queue submit.
+  for (uint32_t submit_idx = 0; submit_idx < submitCount; submit_idx++) {
+    // Go through all command buffers in each submit.
+    for (uint32_t cmd_buffer_idx = 0;
+         // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+         cmd_buffer_idx < pSubmits[submit_idx].commandBufferCount;
+         cmd_buffer_idx++) {
+      auto* command_buffer_handle =
+          // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+          pSubmits[submit_idx].pCommandBuffers[cmd_buffer_idx];
+
+      auto* command_buffer =
+          device_data->command_buffers.get(command_buffer_handle);
+
+      // Ignore command buffers that are not tracked, i.e. command buffers that
+      // don't contain any interesting commands.
+      if (command_buffer == nullptr) {
+        continue;
+      }
+      // Mark the command buffer as submitted.
+      command_buffer->SetSubmitted();
+
+      // Skip all command buffers that don't contain any draw calls.
+      if (!command_buffer->ContainsDrawCalls()) {
+        continue;
+      }
+
+      // Create a new draw call state tracker. Tracker stores the state of the
+      // command buffer being processed, i.e. what pipeline is bound, push
+      // constant values, bound vertex and index buffers, etc. This information
+      // is needed when a draw call needs to be processed to an Amber file.
+      DrawCallStateTracker draw_call_state = {};
+      draw_call_state.command_buffer_handle = command_buffer_handle;
+      draw_call_state.queue = queue;
+
+      for (const auto& cmd : *command_buffer->GetCommandList()) {
+        if (auto* cmd_begin_renderpass = cmd->AsBeginRenderPass()) {
+          draw_call_state.current_render_pass =
+              &cmd_begin_renderpass->render_pass_begin;
+          draw_call_state.current_subpass = 0;
+        } else if (auto* cmd_draw = cmd->AsDraw()) {
+          HandleDrawCall(draw_call_state, *device_data, 0, 0,
+                         cmd_draw->first_vertex, cmd_draw->vertex_count,
+                         cmd_draw->first_instance, cmd_draw->instance_count);
+        } else if (auto* cmd_draw_indexed = cmd->AsDrawIndexed()) {
+          HandleDrawCall(
+              draw_call_state, *device_data, cmd_draw_indexed->first_index,
+              cmd_draw_indexed->index_count, 0, 0,
+              cmd_draw_indexed->first_index, cmd_draw_indexed->instance_count);
+        } else if (auto* cmd_bind_pipeline = cmd->AsBindPipeline()) {
+          // Currently we are interested in graphics pipelines only.
+          if (cmd_bind_pipeline->pipeline_bind_point ==
+              VK_PIPELINE_BIND_POINT_GRAPHICS) {
+            draw_call_state.graphics_pipeline = cmd_bind_pipeline->pipeline;
+          }
```

Comment:
I quite like this because you can see all the cases, and it does feel like the code is quite "local" to this scope. However:

1. We could just use RTTI (`dynamic_cast`) instead of using a home-grown `AsX()` approach.
2. Use of RTTI (or any home-grown approach) is strongly discouraged by the Google C++ Style guide.
3. We could have a virtual method on Cmd called something like `ProcessSubmittedCommand` where we could put this code without needing to use RTTI. This last approach would also allow us to make all member variables private.

Let's discuss in tomorrow's meeting. I don't really mind either way, but we may as well discuss since we are so close to the meeting.